server/channels/app/import_functions.go

```diff
@@ -1272,10 +1273,60 @@ func (a *App) importReplies(rctx request.CTX, data []imports.ReplyImportData, po
 	return nil
 }
 
+func compareFilesContent(fileA, fileB io.Reader, bufSize int64) (bool, error) {
+	aHash := sha1.New()
+	bHash := sha1.New()
+
+	if bufSize == 0 {
+		bufSize = 1024 * 1024 * 2 // 2MB
+	}
+
+	var nA, nB int64
+	var errA, errB error
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		var buf []byte
+		// If the reader has a WriteTo method (e.g. *os.File)
+		// we can avoid the buffer allocation.
+		if _, ok := fileA.(io.WriterTo); !ok {
+			buf = make([]byte, bufSize)
+		}
+		nA, errA = io.CopyBuffer(aHash, fileA, buf)
+	}()
+	go func() {
+		defer wg.Done()
+		var buf []byte
+		// If the reader has a WriteTo method (e.g. *os.File)
+		// we can avoid the buffer allocation.
+		if _, ok := fileA.(io.WriterTo); !ok {
+			buf = make([]byte, bufSize)
+		}
+		nB, errB = io.CopyBuffer(bHash, fileB, buf)
+	}()
+	wg.Wait()
+
+	if errA != nil {
+		return false, fmt.Errorf("failed to compare files: %w", errA)
+	}
+
+	if errB != nil {
+		return false, fmt.Errorf("failed to compare files: %w", errB)
+	}
+
+	if nA != nB {
+		return false, fmt.Errorf("size mismatch: %d != %d", nA, nB)
+	}
+
+	return bytes.Equal(aHash.Sum(nil), bHash.Sum(nil)), nil
```

Comment:
I feel that we could probably have a custom implementation that checks each chunk from both files and bails out early if any of them differ. This avoids having to hash the whole file if only to find out that bytes differ from 10th byte onwards. But anyways this is not the primary use-case as it gets introduced only for the second run. We can revisit if this does become a bottleneck later.

Reply:
Yeah but to be honest I am not even sure if we need to go as far as checking shasums in the first place. Both file path and file sizes are the same at this point so we are essentially checking against some kind of corruption (unlikely I'd say). Not exactly sure why it was introduced as it sounds more valuable as an import validation step.

Reply:
@agnivade Still had the benchmark instance up so gave it a quick try. And interestingly enough SHA256 is looking like outperforming SHA1.


*sha1*

```
goos: linux
goarch: amd64
pkg: github.com/mattermost/mattermost/server/v8/channels/app
cpu: Intel(R) Xeon(R) Platinum 8488C
BenchmarkCompareFilesContent/plain/local-8   	     100	1173016844 ns/op	   66128 B/op	      21 allocs/op
PASS
Dropped temporary database dbiukj7o7onjfxxg5omfbxamef3h
ok  	github.com/mattermost/mattermost/server/v8/channels/app	119.388s
```

*sha256*

```
goos: linux
goarch: amd64
pkg: github.com/mattermost/mattermost/server/v8/channels/app
cpu: Intel(R) Xeon(R) Platinum 8488C
BenchmarkCompareFilesContent/plain/local-8   	     100	 741355493 ns/op	   66208 B/op	      21 allocs/op
PASS
Dropped temporary database db1b83ykofftyrucf79oryr4boue
ok  	github.com/mattermost/mattermost/server/v8/channels/app	75.808s
```

Reply:
Updated

Reply:
Why are you surprised? :P It's more widely used, so I'd expect it to be more optimized.